File: program_stack/algorithms/sorting/heap_sort/java/HeapSort.java
import java.util.Arrays;

public class HeapSort {
    /** In-place heapsort (non-decreasing). Not stable. */
    public static void sort(int[] a) {
        if (a == null || a.length < 2) {
            return;
        }
        int n = a.length;
        for (int i = n / 2 - 1; i >= 0; i--) {
            siftDown(a, n, i);
        }
        for (int end = n - 1; end > 0; end--) {
            int tmp = a[0];
            a[0] = a[end];
            a[end] = tmp;
            siftDown(a, end, 0);
        }
    }

    private static void siftDown(int[] a, int heapSize, int i) {
        while (true) {
            int largest = i;
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            if (left < heapSize && a[left] > a[largest]) {
                largest = left;
            }
            if (right < heapSize && a[right] > a[largest]) {
                largest = right;
            }
            if (largest == i) {
                return;
            }
            int tmp = a[i];
            a[i] = a[largest];
            a[largest] = tmp;
            i = largest;
        }
    }

    private static void expect(int[] input, int[] want) {
        int[] a = Arrays.copyOf(input, input.length);
        sort(a);
        if (!Arrays.equals(a, want)) {
            System.err.println("got " + Arrays.toString(a) + " want " + Arrays.toString(want));
            System.exit(1);
        }
    }

    public static void main(String[] args) {
        expect(new int[] {}, new int[] {});
        expect(new int[] {42}, new int[] {42});
        expect(new int[] {1, 2, 3}, new int[] {1, 2, 3});
        expect(new int[] {3, 2, 1}, new int[] {1, 2, 3});
        expect(new int[] {5, 1, 4, 2, 8}, new int[] {1, 2, 4, 5, 8});
        expect(new int[] {3, 1, 2, 1, 3}, new int[] {1, 1, 2, 3, 3});
        System.out.println("heap_sort: ok");
    }
}

File: program_stack/algorithms/sorting/heap_sort/cpp/heap_sort.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <vector>

static void sift_down(std::vector<int>& a, std::size_t heap_size, std::size_t i) {
    while (true) {
        std::size_t largest = i;
        std::size_t left = 2 * i + 1;
        std::size_t right = 2 * i + 2;
        if (left < heap_size && a[left] > a[largest]) {
            largest = left;
        }
        if (right < heap_size && a[right] > a[largest]) {
            largest = right;
        }
        if (largest == i) {
            return;
        }
        std::swap(a[i], a[largest]);
        i = largest;
    }
}

// In-place heapsort (non-decreasing). Not stable.
void sort(std::vector<int>& a) {
    if (a.size() < 2) {
        return;
    }
    std::size_t n = a.size();
    for (std::size_t i = n / 2; i-- > 0;) {
        sift_down(a, n, i);
    }
    for (std::size_t end = n - 1; end > 0; --end) {
        std::swap(a[0], a[end]);
        sift_down(a, end, 0);
    }
}

static void expect(std::vector<int> input, const std::vector<int>& want) {
    sort(input);
    if (input != want) {
        std::cerr << "mismatch\n";
        std::exit(1);
    }
}

int main() {
    expect({}, {});
    expect({42}, {42});
    expect({1, 2, 3}, {1, 2, 3});
    expect({3, 2, 1}, {1, 2, 3});
    expect({5, 1, 4, 2, 8}, {1, 2, 4, 5, 8});
    expect({3, 1, 2, 1, 3}, {1, 1, 2, 3, 3});
    std::cout << "heap_sort: ok\n";
    return 0;
}
